iDiff/differs/differs.go

```diff
@@ -1,86 +1,50 @@
 package differs
 
 import (
-	"bytes"
 	"errors"
-	"os"
-	"reflect"
 
 	"github.com/GoogleCloudPlatform/runtimes-common/iDiff/utils"
 )
 
-var diffs = map[string]func(string, string, bool, bool) (string, error){
-	"hist":    HistoryDiff,
-	"history": HistoryDiff,
-	"file":    FileDiff,
-	"apt":     AptDiff,
-	"linux":   AptDiff,
-	"pip":     PipDiff,
-	"node":    NodeDiff,
+type ImageDiff struct {
+	Image1    utils.Image
+	Image2    utils.Image
+	DiffType  Differ
+	UseDocker bool
 }
 
-func Diff(arg1, arg2, differ string, json bool, eng bool) (string, error) {
-	if f, exists := diffs[differ]; exists {
-		fValue := reflect.ValueOf(f)
-		histValue := reflect.ValueOf(HistoryDiff)
-		fileValue := reflect.ValueOf(FileDiff)
-		if fValue.Pointer() == histValue.Pointer() || fValue.Pointer() == fileValue.Pointer() {
-			return f(arg1, arg2, json, eng)
-		}
-		return specificDiffer(f, arg1, arg2, json, eng)
-	}
-	return "", errors.New("Unknown differ")
+type DiffResult interface {
+	OutputJSON() error
+	OutputText() error
 }
 
-func specificDiffer(f func(string, string, bool, bool) (string, error), img1, img2 string, json bool, eng bool) (string, error) {
-	var buffer bytes.Buffer
-	validDiff := true
-	imgPath1, err := utils.ImageToFS(img1, eng)
-	if err != nil {
-		buffer.WriteString(err.Error())
-		validDiff = false
-	}
-	imgPath2, err := utils.ImageToFS(img2, eng)
-	if err != nil {
-		buffer.WriteString(err.Error())
-		validDiff = false
-	}
-
-	var diff string
-	if validDiff {
-		output, err := f(imgPath1, imgPath2, json, eng)
-		if err != nil {
-			buffer.WriteString(err.Error())
-		}
-		diff = output
-	}
-
-	errStr := remove(imgPath1, true)
-	errStr += remove(imgPath2, true)
-	if errStr != "" {
-		buffer.WriteString(errStr)
-	}
+type Differ interface {
+	Diff(image1, image2 utils.Image, eng bool) (DiffResult, error)
+}
 
-	if buffer.String() != "" {
-		return diff, errors.New(buffer.String())
-	}
-	return diff, nil
+var diffs = map[string]Differ{
+	"hist":    HistoryDiffer{},
+	"history": HistoryDiffer{},
+	"file":    FileDiffer{},
+	"apt":     AptDiffer{},
+	"linux":   AptDiffer{},
+	"pip":     PipDiffer{},
+	"node":    NodeDiffer{},
 }
 
-func remove(path string, dir bool) string {
-	var errStr string
-	if path == "" {
-		return ""
-	}
+func (diff ImageDiff) GetDiff() (DiffResult, error) {
+	img1 := diff.Image1
+	img2 := diff.Image2
+	differ := diff.DiffType
+	eng := diff.UseDocker
```

Comment:
yeah honestly it was only included there because that's where all the other args (the two images and differ) are collected from the command line, but this eng bool that's currently required in a lot of our functions is in the process of being removed because @abbytiz is looking into Docker Registry stuff as per Dan's suggestion to remove Docker dependencies entirely... but in the meantime, should I go ahead and do that?